.clj-kondo/hooks/clojure/core.clj

```diff
@@ -179,6 +179,136 @@
+  (into global-modules internal-modules))
+
+(defn- module->child-detector
+  "Construct a function that return the given namespace when called with a namespace which is nested inside it."
+  [module-ns-sym]
+  (let [prefix (str (name module-ns-sym) ".")]
+    (fn child-detector [ns-sym]
+      (when (str/starts-with? (name ns-sym) prefix)
+        module-ns-sym))))
+
+(def ^:private module-detectors
+  "A pre-built list of detectors corresponding to our modules, going from most to least specific."
+  (->> (reverse modules)
+       (mapv module->child-detector)))
+
+(defn- parent-module
+  "Find the most tightly enclosing module for the given namespace, returning nil if it is not within any."
+  [ns-sym]
+  (some #(% ns-sym) module-detectors))
+
+(defn- base-module
+  "Traverse up through the module hierarchy to find the top-level module for the given namespace, if it is within one."
+  [ns-sym]
+  (when-let [parent (parent-module ns-sym)]
+    (loop [module parent]
+      (if-let [parent (parent-module module)]
+        (recur parent)
+        module))))
+
+(when-not (every? (comp global-modules base-module) internal-modules)
+  (throw (Exception. "Sanity check: every internal module is inside a global module")))
+
+(defn- dependency-info
+  "Return the metadata corresponding to each namespace / class within a given section of an ns form."
+  [node]
+  (if-let [ns-sym (:value node)]
+    ;; It could be a naked symbol
+    [{:node node, :ns-sym ns-sym}]
+    (let [[{base :value :as first-node} & sub-nodes] (:children node)]
+      (if-not (symbol? (:value (first sub-nodes)))
+        ;; Base is either the only form, or followed by a keyword like :as or :refer
+        ;; In either case, it should already be fully qualified.
+        [{:node first-node, :ns-sym base}]
+        ;; If there is a list of symbols, the first node gives the namespace.
+        (for [sub-node sub-nodes]
+          {:node sub-node, :ns-sym (symbol (str (name base) "."
+                                                (name (:value sub-node))))})))))
+
+(def ^:private dependency-section? #{:require :import :use :load})
+
+(defn- dependencies->ns-symbols
+  "Given a subsection of an ns-form, e.g. (:require ...), return a list of info for the namespaces it references."
+  [dependency-node]
+  (let [[{section-key :k} & references] (:children dependency-node)]
+    (when (dependency-section? section-key)
+      (mapcat dependency-info references))))
+
+(defn- remove-suffix
+  "If the given string has the given suffix, return it with the suffix removed."
+  [s suffix]
+  (when (str/ends-with? s suffix)
+    (subs s 0 (- (count s)
+                 (count suffix)))))
+
+(defn- allowed-parents
+  "The modules whose internals the given namespace is allowed to depend on."
+  [{ns-sym :value ns-s :string-value}]
+  (into #{} (remove nil?)
+        [;; Modules can depend on their own internals.
+         (modules ns-sym)
+         ;; We can depend on our siblings.
```

Comment:
It feels a bit weird for modules to be able to depend on sibling modules. We may want to tighten this up a bit.